### ISSUE TITLE
BUG: np.vectorize doesn't work inside class definition

### ISSUE THREAD

**Issue by daniilS** [numpy/numpy#24397]
### Describe the issue:

`np.vectorize` doesn't work when used on a method inside a class definition, whether it's used as a function or as a decorator. The docstring for `vectorize` says that `pyfunc` is "A python function or method".

#9349 was working on a fix by making `vectorize` follow the descriptor protocol, but was closed when #23514 made `np.vectorize` work as a decorator. However, #23514 didn't implement the `vectorize.__get__` method which would fix this.

### Reproduce the code example:

```python
import numpy as np

class C:
    def __init__(self, y):
        self.y = y
        self.inside_init = np.vectorize(self.pyfunc)  # current workaround
    
    @np.vectorize
    def decorated(self, x):
        return x + self.y
    
    def pyfunc(self, x):
        return x + self.y
    
    inside_class = np.vectorize(pyfunc)

c = C(1)
a = np.array([1, 2])

c.inside_init(a)  # works
c.decorated(a)  # error
c.inside_class(a)  # error
```

### Error message:

```shell
[...]
TypeError: C.decorated() missing 1 required positional argument: 'x'
[...]
TypeError: C.pyfunc() missing 1 required positional argument: 'x'
```

### Runtime information:

1.25.2
3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 64 bit (AMD64)]
[{'numpy_version': '1.25.2',
  'python': '3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 '
            '64 bit (AMD64)]',
  'uname': uname_result(system='Windows', node='LAPTOP-11P3HCJO', release='10', version='10.0.19045', machine='AMD64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}}]

**Comment by rkern** [comment]
`inside_class = np.vectorize(pyfunc)` works to vectorize the function `def pyfunc(self, x):`, but at that particular point in the execution, note that it has two arguments, `self` and `x`. It's exactly that function, no more or less. It's not a method yet until the class definition is complete and Python assembles the actual class object at the end.

The "workaround" is the one correct way to implement this. The decorator won't (and can't) work on the method definition inside a class, and that isn't what was intended to be described by the docstring. When it said "method" there it meant an actual `instancemethod` as in the workaround case. This is a documentation issue.

**Comment by rkern** [comment]
No, you're right, but it's an enhancement. The error was just that gh-23514 did not in fact supercede gh-9349.
